agent/drive_handler.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+)
+
+type drive struct {
+	Name       string
+	DriveID    string
+	MajorMinor string
+	Size       string
+	Type       string
+}
+
+type driveHandler struct {
+	drives []drive
+}
+
+// discoverDrives will use lsblk to populate the drives field in
+// the drive handler.
+func (d *driveHandler) discoverDrives() error {
+	drives, err := lsblk()
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(drives); i++ {
+		if !isStubDrive(drives[i]) {
+			continue
+		}
+
+		id := getDriveID(drives[i])
+		drives[i].DriveID = id
+	}
+
+	fmt.Println(drives)
+	d.drives = drives
+	return nil
+}
+
+// lsblk will call the lsblk linux command and parse the given output into an
+// array of devices.
+func lsblk() ([]drive, error) {
+	cmd := exec.Command("lsblk")
+	b, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+
+	reader := bytes.NewReader(b)
+	drives, err := parseLsblkOutput(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	return drives, nil
+}
+
+// parseLsblkOutput will parse the given output from the lsblk call.
+func parseLsblkOutput(reader io.Reader) ([]drive, error) {
+	scanner := bufio.NewScanner(reader)
+	splitFn := func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		advance, token, err = bufio.ScanLines(data, atEOF)
+		return
+	}
+
+	scanner.Split(splitFn)
+
+	drives := []drive{}
+	atFirstRow := true
+	for scanner.Scan() {
+		text := scanner.Text()
+		columns := []string{}
+		for len(text) > 0 {
+			advance, token, err := bufio.ScanWords([]byte(text), len(text) > 0)
+			if err != nil {
+				return nil, err
+			}
+
+			columns = append(columns, string(token))
+			text = text[advance:]
+		}
+
+		if !atFirstRow {
```

Comment:
This parsing logic is quite complex. How do you feel about using `--json` and `--output` flags, so parsing reduces to `json.Unmarshal` + model def?

For instance:
```
$ lsblk --json --output name,size,type,maj:min --nodeps
{
   "blockdevices": [
      {"name": "loop0", "size": "100M", "type": "loop", "maj:min": "7:0"},
      {"name": "sda", "size": "931.5G", "type": "disk", "maj:min": "8:0"},
      {"name": "nvme0n1", "size": "477G", "type": "disk", "maj:min": "259:0"}
   ]
}
```

Reply:
@mxpv - Yes! I love the fact there's a `--json`. Will update.